File: src/gc/david/dfm/Haversine.java
package gc.david.dfm;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Locale;

/**
 * Haversine class defines static methods to calculate distances between two
 * points on a sphere, in this case, the Earth, from their longitudes and
 * latitudes.
 * 
 * @author David
 * @see <a href="http://en.wikipedia.org/wiki/Haversine_formula">Haversine</a>
 * 
 */
public class Haversine {

	/**
	 * Earth radio
	 */
	private static final double RADIUS = 6371000;

	/**
	 * Calculate distance between two positions in meters using JNI.
	 * 
	 * @param lat_a
	 *            Current position latitude in degrees.
	 * @param lon_a
	 *            Current position longitude in degrees.
	 * @param lat_b
	 *            Destination position latitude in degrees.
	 * @param lon_b
	 *            Destination position longitude in degrees.
	 * @return Distance in meters.
	 */
	public static native double getDistanceJNI(double lat_a, double lon_a,
			double lat_b, double lon_b);

	/**
	 * Calculate distance between two positions in meters.
	 * 
	 * @param lat_a
	 *            Current position latitude in degrees.
	 * @param lon_a
	 *            Current position longitude in degrees.
	 * @param lat_b
	 *            Destination position latitude in degrees.
	 * @param lon_b
	 *            Destination position longitude in degrees.
	 * @return Distance in meters.
	 */
	public static double getDistance(double lat_a, double lon_a, double lat_b,
			double lon_b) {
		double lat1 = Math.toRadians(lat_a);
		double lat2 = Math.toRadians(lat_b);
		double lon1 = Math.toRadians(lon_a);
		double lon2 = Math.toRadians(lon_b);

		double dLat = lat2 - lat1;
		double dLon = lon2 - lon1;

		double sinlat = Math.sin(dLat / 2);
		double sinlon = Math.sin(dLon / 2);

		double a = (sinlat * sinlat) + Math.cos(lat1) * Math.cos(lat2)
				* (sinlon * sinlon);
		double c = 2 * Math.asin(Math.min(1.0, Math.sqrt(a)));

		return RADIUS * c;
	}

	/**
	 * Give the same format to all distances.
	 * 
	 * @param d
	 *            Distance in meters without format.
	 * @return Formatted distance.
	 */
	private static double daleFormato(double d) {
		// Estudiar la posibilidad de diferenciar la cantidad atendiendo al
		// Locale
		DecimalFormatSymbols simbolos = new DecimalFormatSymbols();
		simbolos.setDecimalSeparator('.');
		DecimalFormat formatter = new DecimalFormat("#####0.00000", simbolos);
		String miNumero = formatter.format(d);

		return Double.valueOf(miNumero);
	}

	/**
	 * Normalize distance corresponding to its unit and the device locale and
	 * with only two decimal digits. This static function distinguish between
	 * metric and imperial and US customary units.
	 * 
	 * @param distance
	 *            Formatted distance.
	 * @param locale
	 *            The current locale of the device.
	 * @return A String with the amount and the unit.
	 */
	public static String normalize(double distance, Locale locale) {

		String resultado;

		double measure;

		DecimalFormat formatter = new DecimalFormat("###,##0.00");

		String unidadMedida = "km";
		if (locale.equals(Locale.CANADA) || locale.equals(Locale.CHINA)
				|| locale.equals(Locale.JAPAN) || locale.equals(Locale.KOREA)
				|| locale.equals(Locale.TAIWAN) || locale.equals(Locale.UK)
				|| locale.equals(Locale.US)) {
			unidadMedida = "mi";
			formatter = new DecimalFormat("###,##0.00");

			if (distance >= 1609.344) // Hay al menos una milla
				measure = distance / 1609.344;
			else { // No llega a una milla, medimos en yardas
				unidadMedida = "yd";
				measure = distance * 1.093613298337708;
			}
		} else {
			if (distance >= 1000.0) // Hay al menos un kilómetro
				measure = distance / 1000;
			else { // No llega a un kilómetro, medimos en metros
				unidadMedida = "m";
				measure = distance;
			}
		}
		resultado = formatter.format(measure);
		return resultado + " " + unidadMedida/* + (measure>1 ? "s" : "") */;
	}

	// Load the JNI library
	static {
		System.loadLibrary("gc_david_dfm_Haversine");
	}
}

File: common/src/main/java/gc/david/dfm/map/Haversine.kt
/*
 * Copyright (c) 2026 David Aguiar Gonzalez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gc.david.dfm.map

import java.text.DecimalFormat
import kotlin.math.*

/**
 * Haversine class implements static methods to calculate distances between two
 * points on a sphere, in this case, the Earth, from their longitudes and
 * latitudes. It also defines static methods to calculate elevation points.
 *
 * @author David
 * @see [Haversine](http://en.wikipedia.org/wiki/Haversine_formula)
 */
object Haversine {

    private const val EARTH_RADIUS_IN_METRES = 6371000.0
    private const val MILE_IN_METRES = 1609.344
    private const val KILOMETRE_IN_METRES = 1000.0
    private const val YARDS_PER_METRE = 1.093613298337708
    private const val FEET_IN_METRES = 0.3048

    /**
     * Calculates distance between two positions in metres.
     *
     * @param latitudeA  Current position latitude in degrees.
     * @param longitudeA Current position longitude in degrees.
     * @param latitudeB  Destination position latitude in degrees.
     * @param longitudeB Destination position longitude in degrees.
     * @return Distance in meters.
     */
    fun getDistance(latitudeA: Double,
                    longitudeA: Double,
                    latitudeB: Double,
                    longitudeB: Double): Double {
        val latitudeAInRadians = Math.toRadians(latitudeA)
        val longitudeAInRadians = Math.toRadians(longitudeA)
        val latitudeBInRadians = Math.toRadians(latitudeB)
        val longitudeBInRadians = Math.toRadians(longitudeB)

        val distanceLatitudes = latitudeBInRadians - latitudeAInRadians
        val distanceLongitudes = longitudeBInRadians - longitudeAInRadians

        val sinLatitude = sin(distanceLatitudes / 2)
        val sinLongitude = sin(distanceLongitudes / 2)

        val a = sinLatitude * sinLatitude + cos(latitudeAInRadians) * cos(latitudeBInRadians) * (sinLongitude * sinLongitude)
        val c = 2 * asin(min(1.0, sqrt(a)))

        return EARTH_RADIUS_IN_METRES * c
    }

    /**
     * Normalizes distance corresponding to its unit system with only two decimal
     * digits. Distinguishes between metric and imperial units.
     *
     * @param distanceInMetres Unformatted distance in metres.
     * @param unitSystem       The unit system to use for formatting.
     * @return A String with the amount and the unit.
     */
    fun normalizeDistance(distanceInMetres: Double, unitSystem: UnitSystem): String {
        val (measureUnit, distance) = when (unitSystem) {
            UnitSystem.IMPERIAL -> if (distanceInMetres >= MILE_IN_METRES) {
                "mi" to distanceInMetres / MILE_IN_METRES
            } else {
                "yd" to distanceInMetres * YARDS_PER_METRE
            }
            UnitSystem.METRIC -> if (distanceInMetres >= KILOMETRE_IN_METRES) {
                "km" to distanceInMetres / KILOMETRE_IN_METRES
            } else {
                "m" to distanceInMetres
            }
        }
        return "${DecimalFormat("##,##0.00").format(distance)} $measureUnit"
    }

    /**
     * Normalizes altitude corresponding to its unit system with only two decimal
     * digits. Distinguishes between metric and imperial units.
     *
     * @param altitude   Unformatted altitude in metres.
     * @param unitSystem The unit system to use for formatting.
     * @return A double with only the normalized amount.
     */
    fun normalizeAltitude(altitude: Double, unitSystem: UnitSystem): Double {
        val measure = when (unitSystem) {
            UnitSystem.IMPERIAL -> altitude / FEET_IN_METRES
            UnitSystem.METRIC -> altitude
        }

        // Two decimal digits
        return (measure * 1e2).roundToLong() / 1e2
    }

    /**
     * Returns a string with the altitude unit, m or ft, based on the unit system.
     *
     * @param unitSystem The unit system to use.
     * @return String with altitude unit.
     */
    fun getAltitudeUnit(unitSystem: UnitSystem): String {
        return when (unitSystem) {
            UnitSystem.IMPERIAL -> "ft"
            UnitSystem.METRIC -> "m"
        }
    }
}
